lambda/src/main/scala/pricemigrationengine/migrations/GW2024Migration.scala

```diff
@@ -48,33 +64,35 @@ object GW2024Migration {
     }
   }
 
-  // ------------------------------------------------
-  // Data Extraction and Manipulation
-  // ------------------------------------------------
+  // Note about the difference between lastChangeTypeIsAdd and lastChangeTypeIsNotRemove
+  // Sadly `lastChangeType` is not always defined on all rate plans. The situation is:
+  //     - Not defined                    -> Active rate plan
+  //     - Defined and value is "Add"     -> Active rate plan
+  //     - Defined and value is "Removed" -> Non active rate plan
+
+  def lastChangeTypeIsAdd(ratePlan: ZuoraRatePlan): Boolean = {
+    ratePlan.lastChangeType.fold(false)(_ == "Add")
+  }
+
+  def lastChangeTypeIsNotRemove(ratePlan: ZuoraRatePlan): Boolean = {
+    ratePlan.lastChangeType.fold(true)(_ != "Remove")
+  }
 
   def subscriptionToMigrationRatePlans(subscription: ZuoraSubscription): List[ZuoraRatePlan] = {
-    val migrationRatePlanNames = List(
-      "GW Oct 18 - Monthly - Domestic",
-      "GW Oct 18 - Quarterly - Domestic",
-      "GW Oct 18 - Annual - Domestic",
-      "GW Oct 18 - Monthly - ROW",
-      "GW Oct 18 - Quarterly - ROW",
-      "GW Oct 18 - Annual - ROW",
-      "Guardian Weekly Quarterly",
-      "Guardian Weekly Annual",
-      "GW Oct 18 - 1 Year - Domestic",
-      "Guardian Weekly 1 Year",
-    )
-    subscription.ratePlans.filter(rp => migrationRatePlanNames.contains(rp.ratePlanName))
+    subscription.ratePlans
+      .filter(rp => lastChangeTypeIsNotRemove(rp))
+      .filter(rp => migrationRatePlanNames.contains(rp.ratePlanName))
   }
 
   def subscriptionToMigrationRatePlan(subscription: ZuoraSubscription): Option[ZuoraRatePlan] = {
-    subscriptionToMigrationRatePlans(subscription: ZuoraSubscription).headOption
+    subscriptionToMigrationRatePlans(subscription) match {
```

Comment:
what's the reason for replacing `headOption` here?

Reply:
@tomrf1 

The code that you asked the question about is a previous version. I have updated the implementation in subsequent commits.  

That having been said, let me answer your question because it's an interesting one. It was a precaution. I wanted to take account of the extremely unlikely but still not impossible situation of a subscription having two rate plans of different names, both names appearing in the list of migration rate plan names and both being active. Something like that would not happen in practice, but could happen in theory. 

In any case, the above is now completely irrelevant, because I downloaded all 33,000+ subscriptions of the migration and tested all of them one by one. They are fine, so a lazier implementation (eg: going back to `headOption`) works perfectly fine.